version.sh

```diff
@@ -0,0 +1,125 @@
+#!/bin/bash
+
+# This script dynamically determines an appropriate version string for the
+# currently checked-out commit. As tag versions will typically be provided by CI
+# for releases, this script is primarily to support local builds that work as
+# one would expect.
+#
+# Consider the following git DAG representing a hypothetical Determined git
+# tree:
+#
+#                     HEAD
+#                      |
+#                  I---J (feat-behind-latest-tag)
+#                /
+#...A---B---C---D---E---F---M---N---O (main)
+#    \                   \
+#     G---H (1.1.x)       K---L (1.2.x)
+#         |                   |
+#       1.1.0               1.2.0
+#
+# The newest version is 1.2.0, as tagged on the release branch 1.2.x, at L. The
+# previous version is 1.1.0, on release branch 1.1.x, at H. Our feature branch
+# is feat-behind-latest-tag, with HEAD at J. The goal is to output the nearest
+# tag on the release branch behind wherever we are. At first, I tried to finagle
+# git-rev-list to make this happen, but I was unable to get a working
+# solution. So instead, I used a combination of git-describe and git-tag.
+#
+# git-describe will give us a tag if we're currently on a release branch, or any
+# branch with tags, which is unlikely, but possible. If we're on a feature
+# branch, which is much more likely, we need to use git-tag and git-merge-base
+# to work backward: we find the merge-base of HEAD and main, then search for all
+# tags that don't contain that commit (i.e. all tags created before that
+# commit). From there, we just sort and filter the tag list, and grab the top
+# element, which is the most recent, previous tag.
+#
+# So, in our diagram, if run from B, C, D, E, F, I, J, or K, the script will return
```

Comment:
So just to be clear - after we release 1.1.0, dev builds for the next couple of months will be called 1.1.0-dev? I think that's probably fine, but we'll need to make sure we communicate that well because devs will be used to the version number immediately being bumped up (which I've never liked, because it makes an assumption about what the next version will be before we've even decided - although now we won't have that specific problem if we follow our plans).

Reply:
Not quite. After we release `1.1.0`, commit-level CI jobs will present with `1.1.0+SHA` as their version internally. I was attempting to emulate `git describe`, which will return the most recent tag, the number of commits since that tag, followed by the current commit hash.

I also wanted each version to properly reflect the uniqueness of the artifacts that get built. Everything after `1.1.0` but before `1.2.0` being `1.1.0-dev` is technically not true, because it maps a single version to many different commits. As far as I can tell, it's not a huge deal because we never build any of those commits in any sort of release capacity anyway (not until the `-rc` versions, and those are explicitly different each time they're released), but I wanted to impose some guarantees on the version specifier.

Reply:
And it can all be overridden locally by setting the `VERSION` environment variable; `version.sh` will use that first if it exists.

Reply:
theoretically, we could have /docs changes on any/all branches, correct?

Reply:
I'm not sure how often Determined, specifically, does it, but for git in general, yes.